pageserver/src/tenant/timeline.rs

```diff
@@ -4155,6 +4161,171 @@ impl Timeline {
+            // partition does not cover any keys owned by this shard.  In this case, to ensure
+            // we don't leave gaps between image layers, leave `start` where it is, so that the next
+            // layer we write will cover the key range that we just scanned.
+            tracing::debug!("no data in range {}-{}", img_range.start, img_range.end);
+            Ok(None)
+        }
+    }
+
+    /// Create an image layer for metadata keys. This function produces one image layer for all metadata
+    /// keys for now. Because metadata keys cannot exceed basebackup size limit, the image layer for it
+    /// would not be too large to fit in a single image layer.
+    #[allow(clippy::too_many_arguments)]
+    async fn create_image_layers_for_metadata_keys(
+        self: &Arc<Self>,
+        partition: &KeySpace,
+        mut image_layer_writer: ImageLayerWriter,
+        lsn: Lsn,
+        ctx: &RequestContext,
+        img_range: Range<Key>,
+        start: &mut Key,
+        mode: ImageLayerCreationMode,
+    ) -> Result<Option<ResidentLayer>, CreateImageLayersError> {
+        // Metadata keys image layer creation.
+        let delta_file_accessed_begin = ctx
+            .vectored_access_delta_file_cnt
+            .load(AtomicOrdering::SeqCst);
+        let data = self
+            .get_vectored_impl(
+                partition.clone(),
+                lsn,
+                ValuesReconstructState::default(),
+                ctx,
+            )
+            .await?;
+        let (data, total_kb_retrieved, total_key_retrieved) = {
+            let mut new_data = BTreeMap::new();
+            let mut total_kb_retrieved = 0;
+            let mut total_key_retrieved = 0;
+            for (k, v) in data {
+                let v = v.map_err(CreateImageLayersError::PageReconstructError)?;
+                total_kb_retrieved += KEY_SIZE + v.len();
+                total_key_retrieved += 1;
+                new_data.insert(k, v);
+            }
+            (new_data, total_kb_retrieved / 1024, total_key_retrieved)
+        };
+        let delta_file_accessed = ctx
+            .vectored_access_delta_file_cnt
+            .load(AtomicOrdering::SeqCst)
+            - delta_file_accessed_begin;
+
+        let trigger_generation = delta_file_accessed >= MAX_AUX_FILE_V2_DELTAS;
+        info!("generate image layers for metadata keys: trigger_generation={trigger_generation}, delta_file_accessed={delta_file_accessed}, total_kb_retrieved={total_kb_retrieved}, total_key_retrieved={total_key_retrieved}");
+        if !trigger_generation && mode == ImageLayerCreationMode::Try {
+            return Ok(None);
+        }
+        let has_keys = !data.is_empty();
+        for (k, v) in data {
+            // Even if the value is empty (deleted), we do not delete it for now until we can ensure vectored get
+            // considers this situation properly.
+            // if v.is_empty() {
+            //     continue;
+            // }
```

Comment:
I don't understand what situation you're referring to. If this is a bug, please create an issue and link it here.

Reply:
I just don't feel safe of directly removing a single key from the image layer because we didn't do this before. https://github.com/neondatabase/neon/pull/7773 might fix that, but I need to create a test case that ensures everything works before uncommenting this.

Reply:
https://github.com/neondatabase/neon/issues/7800